File: src/public_holidays.py
"""Ontario public holidays for schedule membership (observed dates)."""

from __future__ import annotations

from datetime import date
from functools import lru_cache

import holidays

# Bylaw "public holidays" → Ontario statutory holidays (not City of Toronto HR extras).
_ON = holidays.country_holidays


@lru_cache(maxsize=64)
def _holidays_for_year(year: int) -> holidays.HolidayBase:
    return _ON('CA', subdiv='ON', years=year, observed=True)


def is_public_holiday(slot: dict[str, int]) -> bool:
    """
  Return whether *slot* falls on an Ontario public holiday (observed date).

  Requires ``year`` in *slot* for correct weekend/substitution rules.
  """
    year = slot.get('year')
    if year is None:
        return False
    try:
        d = date(int(year), int(slot['month']), int(slot['dayOfMonth']))
    except (TypeError, ValueError):
        return False
    return d in _holidays_for_year(int(year))

File: web/src/lib/schedule/publicHolidays.ts
import Holidays from 'date-holidays'
import type { Slot } from './types'

const cache = new Map<number, Holidays>()

function holidaysForYear(year: number): Holidays {
  let hd = cache.get(year)
  if (!hd) {
    hd = new Holidays('CA', 'ON')
    cache.set(year, hd)
  }
  return hd
}

/** Ontario statutory public holidays (observed when applicable). */
export function isPublicHoliday(slot: Slot): boolean {
  if (slot.year == null) return false
  const result = holidaysForYear(slot.year).isHoliday(
    new Date(slot.year, slot.month - 1, slot.dayOfMonth),
  )
  if (!result) return false
  const list = Array.isArray(result) ? result : [result]
  return list.some((h) => h.type === 'public')
}
